default/python/AI/SaveGameManager.py

```diff
@@ -0,0 +1,260 @@
+SET_PREFIX = '__SET__'
+TUPLE_PREFIX = '__TUPLE__'
+
+
+# placeholder char to represent quotes in nested containers
+# which would break json decoding if present.
+PLACEHOLDER = chr(26)
+
+
+class CanNotSaveGameException(Exception):
+    """Exception raised when constructing the savegame string failed."""
+    pass
+
+
+class InvalidSaveGameException(Exception):
+    """Exception raised if the savegame could not be loaded."""
+    pass
+
+
+class FreeOrionAISaveGameEncoder(object):
+    """Encoder support for some classes not supported by json."""
+
+    def encode(self, o):
+        if isinstance(o, basestring):
+            pass
+        elif o is None:
+            o = NONE
+        elif o is True:
+            o = TRUE
+        elif o is False:
+            o = FALSE
+        elif isinstance(o, (int, long)):
+            o = INT_PREFIX + str(o)
+        elif isinstance(o, float):
+            o = FLOAT_PREFIX + repr(o)
+        elif isinstance(o, dict):
+            return self._dict_encoder(o)
+        elif isinstance(o, list):
+            return self._list_encoder(o)
+        elif isinstance(o, tuple):
+            o = self._tuple_encoder(o)
+        elif isinstance(o, set):
+            o = self._set_encoder(o)
+        else:
+            return self._object_encoder(o)
+        return "\"%s\"" % o
+
+    def _object_encoder(self, obj):
+        try:
+            if obj.__class__.__module__ not in _trusted_modules:
+                raise CanNotSaveGameException("Module %s is not trusted" % obj.__class__.__module__)
+        except AttributeError:
+            # obj does not have a class or class has no module
+            raise CanNotSaveGameException("Encountered unsupported object %s (%s)" % (obj, type(obj)))
+
+        key = CLASS_PREFIX + obj.__class__.__module__ + '.' + obj.__class__.__name__
+
+        try:
+            getstate = obj.__getstate__
+        except AttributeError:
+            value = obj.__dict__
+        else:
+            # only call now to not incorrectly catch exceptions during getstate call
+            value = getstate()
+        return self._dict_encoder({key: value})
+
+    def _list_encoder(self, o):
+        return '[' + ', '.join(self.encode(v) for v in o) + ']'
+
+    def _tuple_encoder(self, o):
+        return (TUPLE_PREFIX + "(" + self._list_encoder(list(o)) + ")").replace('\"', PLACEHOLDER)
+
+    def _set_encoder(self, o):
+        return (SET_PREFIX + "(" + self._list_encoder(list(o)) + ")").replace('\"', PLACEHOLDER)
+
+    def _dict_encoder(self, o):
+        return '{' + ', '.join('%s: %s' % (self.encode(k), self.encode(v)) for k, v in o.iteritems()) + '}'
+
+
+class FreeOrionAISaveGameDecoder(json.JSONDecoder):
+
+    def __init__(self, **kwargs):
+        super(FreeOrionAISaveGameDecoder, self).__init__(object_hook=self.__parse_savegame, strict=False, **kwargs)
+
+    def decode(self, s, _w=None):
+        return self.__interpret(super(FreeOrionAISaveGameDecoder, self).decode(s))
```

Comment:
I'm concerned about this line... Seeming you're still calling the base class JSONDecoder with an untrusted string. Even if you make the parsing of the resulting dict relatively safe, are you sure it won't attempt to construct an arbitrary class specified in `s` in this call?

Reply:
The default JSONDecoder does not try to construct arbitrary objects but only those specified in the table here: https://docs.python.org/2/library/json.html#encoders-and-decoders

It seems to be safe in regard to arbitrary code execution - at least there are no well known exploits or risks. The only security issue I found in regard to json is to use it to read arbitrary memory (https://bugs.python.org/issue21529) which is fixed in 2.7.5 and later. Regardless, that exploit requires executions of specific commands.

The one issue that remains is that malicious data could still be supplied. So we still have to make sure that the rest of the AI code is not susceptible to arbitrary code execution (e.g. by using ```eval``` on some string that could be (indirectly) supplied by the savegame).